### ISSUE TITLE
Triangle triangle intersect

### ISSUE THREAD

**Issue by Tomas-Ha** [dilevin/computer-graphics-bounding-volume-hierarchy#73]
Hello, currently the brute force method finds significantly more triangles than my tree method. First, #33 states that it was a bonus since it was not covered in class, is this still the case? It was covered in tutorial briefly. 

Second, why might I be experiencing so many missing intersection pairs in my tree? I followed the provided algorithm from the README (https://github.com/dilevin/computer-graphics-bounding-volume-hierarchy/blob/master/README.tex.md#intersection-queries-between-two-trees)

**Comment by Zhecheng-Wang** [comment]
> Hello, currently the brute force method finds significantly more triangles than my tree method. First, #33 states that it was a bonus since it was not covered in class, is this still the case? It was covered in tutorial briefly.

I am not aware of `triangle_triangle_intersection.cpp` being a bonus this year.


> Second, why might I be experiencing so many missing intersection pairs in my tree? I followed the provided algorithm from the README (https://github.com/dilevin/computer-graphics-bounding-volume-hierarchy/blob/master/README.tex.md#intersection-queries-between-two-trees)

Can you be more specific? Does it happen for simple shapes like a torus? What did [diff_and_warn](https://github.com/dilevin/computer-graphics-bounding-volume-hierarchy/blob/183f951e452f9521e6fe1bf2b7f11e6878327698/intersections.cpp#L112) show?

**Comment by Tomas-Ha** [comment]
> Can you be more specific? Does it happen for simple shapes like a torus? What did [diff_and_warn](https://github.com/dilevin/computer-graphics-bounding-volume-hierarchy/blob/183f951e452f9521e6fe1bf2b7f11e6878327698/intersections.cpp#L112) show?

This is an out.txt file of the output of my code. It returns a long list of inputs not found by the tree and then a list of 3 intersections not found by brute force. Is it possible that my triangle algorithm or box algorithm is wrong? I am fairly confident that my find all intersection pairs works.

[out.txt](https://github.com/dilevin/computer-graphics-bounding-volume-hierarchy/files/12857277/out.txt)

I am using the following function call:
`./intersections ../data/knight.obj ../data/cheburashka.obj >> out.txt`

**Comment by Zhecheng-Wang** [comment]
> Is it possible that my triangle algorithm or box algorithm is wrong? I am fairly confident that my find all intersection pairs works.

Could be. I would compare leaf pairs from broad phase with the brute force pairs (so skip narrow phase for now). Basically broad phase pairs is a list of overlapping boxes and it is not likely to get box-box intersection wrong -- so if broad phase box-box pairs is similar to brute force pairs, than there is a bug with your triangle-triangle intersection in narrow phase.

**Comment by Tomas-Ha** [comment]
Could you elaborate on that, I don't fully understand broad phase vs narrow phase and how that relates to my code

**Comment by Tomas-Ha** [comment]
> Could be. I would compare leaf pairs from broad phase with the brute force pairs (so skip narrow phase for now). Basically broad phase pairs is a list of overlapping boxes and it is not likely to get box-box intersection wrong -- so if broad phase box-box pairs is similar to brute force pairs, than there is a bug with your triangle-triangle intersection in narrow phase.

I looked at the README and I think I understand. The broad phase uses the box box and the tree to narrow it down to pairs, so if I have a bunch of pairs being found to overlap that aren't in both cases, it may be my triangle triangle intersection. What formula should I use for that. I looked into Tomas Möller's triangle triangle intersection equation (as mentioned in #12), is there one in the textbook I should be using instead?

**Comment by Zhecheng-Wang** [comment]
> The broad phase uses the box box and the tree to narrow it down to pairs, so if I have a bunch of pairs being found to overlap that aren't in both cases, it may be my triangle triangle intersection.

Yes, if you get AABB and box-box right, box-box pairs from the broad phase should be a superset of triangle-triangle pairs from the narrow phase (and brute force).

> What formula should I use for that. I looked into Tomas Möller's triangle triangle intersection equation (as mentioned in https://github.com/dilevin/computer-graphics-bounding-volume-hierarchy/issues/12), is there one in the textbook I should be using instead?

Hmm, I don't know any from the textbook, as #12 suggested, feel free to implement anything.

**Comment by SCOU7** [comment]
For me what worked is 
1) try adding some print statements in intersections.cpp and check out the size of bf_fairs, leaf_pairs, and tree_pairs. It really helped me for debugging. 
2) in find_all_intersecting_pairs_using_AABBTrees.cpp, whenever you make a pair and add it a collection, the node from rootA must go first and the node from rootB must go second. This is the assumption made in intersections.cpp. This was what made me trouble. 

Hope this helps :)

**Comment by destefy** [comment]
Hey I know this is a year old, but thank you. point 2 saved me

**Comment by SCOU7** [comment]
> Hey I know this is a year old, but thank you. point 2 saved me

I'm happy it helped! Looks like Professor Levin was ahead of the curve. Funny how GitHub issues keep sharing useful tips, even after all this time!
